### ISSUE TITLE
Restore is not checking token validity

### ISSUE THREAD

**Issue by deeksy** [auth0-community/ember-simple-auth-auth0#75]
Currently, when using this plugin, the authorisation flow sets up token expiry in the application through looking at the url hash and logging the user in.  Unfortunately, if a user navigates away from the app, and before coming back to the app the token expires, the restore method of the authorizer doesn't check whether the token is still valid, and because token expiration events are being done in the application through the sessionAuthenticated hook, the app will continue with the token never expiring.

A better approach would be to set up token expiration events in the authenticator, and then call them from both authorise and restore to ensure token expiry is correctly checked on restore.

**Comment by XaserAcheron** [comment]
I'm making a note of this issue, as I had to implement a workaround for this in my own app. Seems like a good idea to put the code where it belongs (in this addon) when I get some free time for tinkering.

**Comment by dja** [comment]
@XaserAcheron Would you mind sharing your workaround, or creating a PR with it? 😁

**Comment by XaserAcheron** [comment]
My app's workaround is pretty blunt; I basically have the following manual check in my application route (paraphrased a bit):
```javascript
	sessionExpired: computed('_expiresAt', function() {
		return getWithDefault(this, '_expiresAt', 0) <= 0;
	}),

	beforeModel() {
		if(get(this, 'session.isAuthenticated')) {
			if(get(this, 'sessionExpired')) {
				// forcibly expire the session
				return this._processSessionExpired();
			} else {
				// do stuff when the user is authenticated
			}
		} else {
			// do stuff when the user is not authenticated
		}
		return this._super(...arguments);
	},
```

This relies on some private stuff that's likely to vanish when this gets fixed for real, so as a fair warning, this will probably break in a future release.

It'll work in the current v3.0.x versions and in the upcoming v3.1.0 release, though, so that's something.

Hope this helps a bit!

**Comment by XaserAcheron** [comment]
Took me long enough, but I've fixed this in `v4.0.1` while I was messing about with token expiration. The token is no longer restored if it's expired, which ought to ease the pain here (and no longer require the above sledgehammer-hack).
